### ISSUE TITLE
Error redrawing 3d points after filtering

### ISSUE THREAD

**Issue by davesteps** [noaa-ocs-hydrography/kluster#91]
This happens intermittently. I am trying to pin down the conditions under which it happens:
 
I subset some data like so:
![image](https://user-images.githubusercontent.com/5738970/161025125-bc7b7c65-4bcb-4e4c-808b-ff95de2721bf.png)

I apply a filter, in this case the angle filter -25+25. When the filter finished the points are redrawn like so:
![image](https://user-images.githubusercontent.com/5738970/161025230-33080795-d137-40e2-94fb-9b4bbdbf744d.png)

However if I reselect the subset, the data is correct, indicating that the data written to disk is correct.
![image](https://user-images.githubusercontent.com/5738970/161025296-e6b7c2c5-71e7-4ed5-b4a3-c2f8ab8f8030.png)

As I said it does not happen all the time. It does seem to be related to having multiple lines selected as I don't think it happens if all the data is from one line.  I will keep trying to get insight into the problem

**Comment by davesteps** [comment]
^ Ignore the point about multiple lines. I have filtered multiple lines without seeing the error

**Comment by ericgyounkin** [comment]
@davesteps is it only on the new dev branch that you see this?  Or did you also see it before.

**Comment by davesteps** [comment]
It can be reproduced with the `0009_20170523_181119_FA2806.all` test data also. Selecting the entire set to view in 3d points view:

**Comment by ericgyounkin** [comment]
oh strange.  OK, that is very helpful, ill take a look at it today.

**Comment by davesteps** [comment]
I've just tried again from scratch:
1) load in the .all file and process
2) build a surface
3) Select all:
![image](https://user-images.githubusercontent.com/5738970/161075135-e96a8912-6220-4d59-8ca4-52c3f7c76603.png)
4) color to rejected:
![image](https://user-images.githubusercontent.com/5738970/161075274-638eb8e0-60b1-49fd-b51c-030da77b088a.png)
5) Filter by angle (use defaults +-45):
![image](https://user-images.githubusercontent.com/5738970/161075430-33500794-eeeb-4f73-b71e-299ac71aa5e7.png)
Results in:
![image](https://user-images.githubusercontent.com/5738970/161075501-5f9d900a-4bff-4bd1-9ef9-5538172f4381.png)
6) redraw subset: 
![image](https://user-images.githubusercontent.com/5738970/161075695-7f8d873d-67df-40df-81f5-e5b4b66160bd.png)

**Comment by davesteps** [comment]
If you do not get the same results then it might be down to my environment. Perhaps with installing our packages something as changed and I'm using different versions to you.

**Comment by ericgyounkin** [comment]
@davesteps Can't reproduce.  But, there is an underlying assumption to the filter -> point view update that could be wrong.  I assume that the order the points are loaded into the points view is the same as the order that the filter results are in, in terms of (time, beam).  This might be a bad assumption, for some reason that you have encountered that i do not understand.  It is certainly a weak assumption, so a fix is probably a good idea regardless.

Either way, I can just throw in another sorting step to match (time,beam) of filter results to points view.  I added a fix to the dev branch.  That should do it.

https://github.com/noaa-ocs-hydrography/kluster/commit/424b27b8976b09fd1220670ea01be931856d9ead

**Comment by ericgyounkin** [comment]
There was one last potential issue I could find, which is where the points view points are not ordered in time, beam.  This should not happen, but just in case, I put in another sort which should align the new status update regardless.

https://github.com/noaa-ocs-hydrography/kluster/commit/3faec83293afb27a283c27b8678cdee630250995

**Comment by davesteps** [comment]
Hi @ericgyounkin, It's strange but I'm still getting it. The fact that you cannot recreate the behaviour at all makes me think it must be something to do with my env. I'll build a new one and test.

**Comment by ericgyounkin** [comment]
@davesteps OK, although I can't think of what would cause the difference.  We have different OS, but I don't see that being the cause.  

I should mention that gridding is going to be probably somewhat broken on the dev branch right now.  I'm in the process of adding/testing the CUBE algorithm in gridding.

**Comment by davesteps** [comment]
Hi @ericgyounkin, so I have started form scratch with a new environment and weirdly still get the behavior. Using Ubuntu 20.04 on `master` (tested on `dev` also) using the steps below to setup and then the processing the data as described above. 

```
conda env create -n kluster_test2 -f environment.yml
conda activate kluster_test2
pip install -e .
mkdir /home/david/miniconda3/envs/kluster_test2/share/qgis/plugins
python -m HSTB.kluster
```

> note the `mkdir` step, it requires that the qgis plugins folder exists even if empty

**Comment by ericgyounkin** [comment]
@davesteps  OK, I updated my ubuntu VirtualBox and I am able to replicate it there.  I can resolve that plugins directory thing easy, it's just in a different place in linux for some reason.  

The filter update bug is so weird.  It only seems to present if you select the whole dataset.  If I get just a slice of the dataset, the update seems to work.  Is that what you are seeing too?

**Comment by davesteps** [comment]
Great that you are seeing it too ;) On the  `0009_20170523_181119_FA2806.all` example yes it is only when the whole data is selected. On the other example it also happens with subsets of data. It is not easy to tell exactly when it happens but if I only have a small subset on one side of line it doesn't tend to happen, however larger subsets over the entire width of the line it is more likely to happen.

**Comment by ericgyounkin** [comment]
@davesteps This is the weirdest thing I think I have come across yet in my programming work.  I resolved the path issues that I found that were linux specific:

https://github.com/noaa-ocs-hydrography/kluster/commit/f8d106a579bedec6c7aadc9e66c0d56a05eb277f

And the issue is gone.  At least in my testing on my ubuntu box.  I don't understand why.  Can you test the new dev branch to make sure I'm not crazy?

**Comment by davesteps** [comment]
Hi @ericgyounkin I tested the latest commit on `dev` and am still seeing it

**Comment by ericgyounkin** [comment]
@davesteps Sorry, this is kind of frustrating.  I still am having issues reproducing on my Ubuntu VM.  I released kluster 0.9.1 yesterday, I was wondering if you could switch to that branch and try.  Might be good to share your 'help - about' screen too, just to ensure the dependency versions match between our environments.

![image](https://user-images.githubusercontent.com/58827984/162205696-ac7ef3e9-a542-4fff-85a4-99652593663b.png)

**Comment by davesteps** [comment]
Hi @ericgyounkin here are my versions:
![image](https://user-images.githubusercontent.com/5738970/162746669-c076dadf-aed9-47d0-8d73-4829c7c876db.png)

**Comment by ericgyounkin** [comment]
Yeah, I really don't understand how that could be.  I was only able to reproduce this once on my virtual box ubuntu instance.  I need to get back to it, I just have a few things I gotta deal with first.  

Do you think you could catch this in a Pycharm debugger?  Or whichever IDE your team uses?  If we could organize another Teams session so that I could see it, maybe that would help.  Just a thought.
